modules/core/shared/src/main/scala/gem/parser/Misc.scala

```diff
@@ -65,5 +65,27 @@ trait MiscParsers {
 
   } namedOpaque s"frac($n)"
 
+
+  /** Force and return all remaining input without consuming it. */
+  val force: Parser[String] =
+    get.flatMap { s =>
+      opt(ensure(s.length + 1)) flatMap {
+        case Some(_) => force
+        case None    => ok(s)
+      }
+    }
+
+  /**
+   * Force and examine remaining input and yield the longest prefix up to (but not including) a
+   * char satisfying `p`, which is discarded. Fails if no such char is found.
+   */
+  def takeUntilLast(p: Char => Boolean): Parser[String] =
+    force.flatMap { s =>
+      s.lastIndexWhere(p) match {
+        case -1 => err[String]("not satisfied")
```

Comment:
What's the type returned by `err[String]`? Is it the same type for all parsing errors in `atto`?

Reply:
It produces a `Parser[String]`, a "Parser that consumes no data and fails with the specified error message." when applied.

Reply:
I've seen in the source and the tutorials that `err` returns a `Failure` which I assume it contains all the context information of the error, like in which line/column or how what's the input remainder (in the case of `takeUntilLast` it doesn't matter). But I can't where `Failure` is defined.

I'm not trying to point at anything wrong here, I'm just trying to learn more about `atto`.

Reply:
Never mind about my previous comment, I [found it](https://github.com/tpolecat/atto/blob/de9e21f5b2ade83102a425089ef9e294fb157cb8/modules/core/src/main/scala/atto/Parser.scala#L75). It seems the error has all the context I was mentioning above. Neat 👍.

Reply:
Error reporting isn't very good. You get an error message (a string), the current stack (which is a list of strings), and the remaining unconsumed input. If you want to know the parse position you can use the `pos` constructor which consumes no input and yields the absolute offset.